tests/std/tests/GH_004109_iter_value_t_direct_initialization/test.compile.pass.cpp

```diff
@@ -0,0 +1,700 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <cassert>
+#include <initializer_list>
+#include <iterator>
+#include <memory>
+#include <numeric>
+#include <random>
+
+#ifdef __cpp_lib_execution
+#include <execution>
+#endif // __cpp_lib_execution
+
+using namespace std;
+
+struct Val;
+
+struct RRef {
+    RRef(const Val&);
+};
+
+struct Val {
+    using difference_type = int;
+    explicit Val(const RRef&);
+    Val& operator=(const RRef&);
+    bool operator==(const Val&) const;
+    bool operator!=(const Val&) const;
+    bool operator<(const Val&) const;
+    bool operator<=(const Val&) const;
+    bool operator>(const Val&) const;
+    bool operator>=(const Val&) const;
+    Val& operator++();
+    Val operator++(int);
+};
+
+struct I {
+    using iterator_category = random_access_iterator_tag;
+    using pointer           = Val*;
+    using value_type        = Val;
+    using reference         = Val&;
+    using difference_type   = int;
+    Val& operator*() const;
+    Val& operator[](int) const;
+    I& operator++();
+    I operator++(int);
+    I& operator--();
+    I operator--(int);
+    I& operator+=(int);
+    I& operator-=(int);
+    friend bool operator==(const I&, const I&);
+    friend bool operator!=(const I&, const I&);
+    friend bool operator<(const I&, const I&);
+    friend bool operator<=(const I&, const I&);
+    friend bool operator>(const I&, const I&);
+    friend bool operator>=(const I&, const I&);
+    friend int operator-(I, I);
+    friend I operator+(I, int);
+    friend I operator-(I, int);
+    friend I operator+(int, I);
+    friend RRef iter_move(const I&);
+};
+
+#ifdef __cpp_lib_concepts
+static_assert(random_access_iterator<I>);
+static_assert(sortable<I>);
+#endif // __cpp_lib_concepts
+
+bool always_true(Val) {
+    return true;
+}
+
+bool always_false(Val) {
+    return false;
+}

```

Comment:
Yes, it's because I started my work from the file https://github.com/microsoft/STL/blob/0403d19f5461fd15983737c3f01ec34800ea9275/tests/std/tests/Dev11_0253803_debug_pointer/test.cpp , but that test is a normal test and not a compile-only test, so some things might be unneeded.

Reply:
Oh, thanks! That explains a lot - we can considerably simplify this test then. I'll work on changes...